### PR TITLE
[data] torch datasource streaming read

### PR DESCRIPTION
**Pull request by Zandew**
## Why are these changes needed?

Provides a parallel and streaming implementation for the `from_torch` read api.

In this PR, we set `parallelism=1`, so we only utilize the streaming portion of this api.

Not included in this PR:
- Parallel read across multiple nodes
- Blocks may be reordered because of parallel read tasks. Add option to preserve order. Not an issue here because we use `parallelism=1`.
- Block metadata, there is no good way of getting the size/schema of a torch dataset. We cannot autodetect parallelism well. Also not a problem for now because of `parallelism=1`

### S3 Images
Image data tak`s3://air-example-data-2/20G-image-data-synthetic-raw/` was read with `parallelism=-1` and iterated over with `iter_torch_batches` on `32vCPU`, `128GiB` nodes.

**6 nodes**
time: ~180s
cpu usage: peak ~56 cores
Screenshot 2023-09-22 at 12 26 58 PM

**1 node**
time: ~240s
cpu usage: peak ~27 cores
Screenshot 2023-09-22 at 12 27 20 PM

Speed up between cases: ~1.33x

The bottleneck here is while deserializing the images afterwards on the driver. Parallelism does not help with this.
Screenshot 2023-09-25 at 3 58 45 PM

### Fake Data
We can generate fake data roughly the same size as the S3 images using `torch.randint(...)` and spinning for some time `for i in range(1e7)` to emulate the scenario above but without the need to deserialize the images.

**6 nodes**
time: ~100s
cpu usage: peak ~170 cores

**1 node**
time: ~550s
cpu usage: peak ~27 cores

Speed up between cases: ~5.5x

## Related issue number

Closes torch read #39287

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/datasource/torch_datasource.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+
+class _TorchDatasourceReader(Reader):
+    def __init__(
+        self, dataset: "torch.utils.data.Dataset", batch_size: int, random_split: bool
+    ):
+        self._dataset = dataset
+        self._batch_size = batch_size
+        self._random_split = random_split
+        self._size = len(cloudpickle.dumps(self._dataset))
```
Comment (scottjlee) [inline python/ray/data/datasource/torch_datasource.py:31]: i'm not sure if this is the best way to calculate dataset size, but i will let others comment since i'm not aware of other size calculation methods for torch Datasets.

File: python/ray/data/datasource/torch_datasource.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        return self._size
+
+
+def _read_subset(subset: "torch.utils.data.Subset", batch_size):
+    import torch
+
+    data_loader = torch.utils.data.DataLoader(
+        subset,
+        # default_collate does not accept `PIL.Image.Image`s
+        collate_fn=lambda x: x,
```
Comment (scottjlee) [inline python/ray/data/datasource/torch_datasource.py:84]: sorry i don't fully follow, can you elaborate on why we need this collate fn?

File: python/ray/data/tests/test_formats.py [hunk 3]
```diff
@@ -185,12 +186,32 @@ def test_from_torch(shutdown_only, tmp_path):
     torch_dataset = torchvision.datasets.MNIST(tmp_path, download=True)
     expected_data = list(torch_dataset)
 
-    ray_dataset = ray.data.from_torch(torch_dataset)
+    ray_dataset = ray.data.from_torch(torch_dataset, parallelism=1)
```
Comment (scottjlee) [inline python/ray/data/tests/test_formats.py:189]: do we need to explicitly specify parallelism here?

File: python/ray/data/tests/test_formats.py [hunk 4]
```diff
@@ -185,12 +186,32 @@ def test_from_torch(shutdown_only, tmp_path):
...
     actual_data = extract_values("item", list(ray_dataset.take_all()))
     assert actual_data == expected_data
 
 
+def test_from_torch_parallel(shutdown_only, tmp_path):
+    # parallel read may reorder blocks, cannot directly assert equality on outputs
+    torch_dataset = torchvision.datasets.MNIST(tmp_path, download=True)
+    expected_data = list(torch_dataset)
+
+    ray_dataset = ray.data.from_torch(torch_dataset, parallelism=3)
```
Comment (scottjlee) [inline python/ray/data/tests/test_formats.py:200]: do we need to explicitly specify parallelism here?

**Review by scottjlee — APPROVED** [review]

File: python/ray/data/datasource/torch_datasource.py (hunk 2 above)
Comment (Zandew) [reply python/ray/data/datasource/torch_datasource.py:84]: I think [default_collate](https://pytorch.org/docs/stable/data.html#torch.utils.data.default_collate) modifies the dataset. Not sure if that is what we want.

File: python/ray/data/tests/test_formats.py (hunk 3 above)
Comment (Zandew) [reply python/ray/data/tests/test_formats.py:189]: This test can fail if we have parallelism > 1 since the block order isn't preserved.

File: python/ray/data/tests/test_formats.py (hunk 4 above)
Comment (Zandew) [reply python/ray/data/tests/test_formats.py:200]: As long as the parallelism autodetection parallelizes when possible,  I guess not

File: python/ray/data/datasource/torch_datasource.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+from ray.data._internal.delegating_block_builder import DelegatingBlockBuilder
+from ray.data.block import BlockMetadata
+from ray.data.datasource.datasource import Datasource, Reader, ReadTask
+
+if TYPE_CHECKING:
+    import torch
+
+
+class TorchDatasource(Datasource):
+    def create_reader(
```
Comment (amogkam) [inline python/ray/data/datasource/torch_datasource.py:15]: can we add docstrings for everything?

File: python/ray/data/datasource/torch_datasource.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+        return read_tasks
+
+    def estimate_inmemory_data_size(self):
+        return self._size
+
+
+def _read_subset(subset: "torch.utils.data.Subset", batch_size):
+    import torch
+
+    data_loader = torch.utils.data.DataLoader(
```
Comment (amogkam) [inline python/ray/data/datasource/torch_datasource.py:81]: do we need DataLoader here? I think we should just access from the underlying Dataset directly.

File: python/ray/data/read_api.py [hunk 7]
```diff
@@ -2282,6 +2283,9 @@ def from_tf(
 @PublicAPI
 def from_torch(
     dataset: "torch.utils.data.Dataset",
+    parallelism: int = -1,
+    random_split: bool = False,
+    batch_size: int = 32,
```
Comment (amogkam) [inline python/ray/data/read_api.py:2288]: Let's add the new args to the docstring?

Also I don't think we want to expose batch_size and random_split? Should it be shuffle instead of random_split?

File: python/ray/data/read_api.py [hunk 8]
```diff
@@ -2313,7 +2317,13 @@ def from_torch(
     Returns:
         A :class:`MaterializedDataset` containing the Torch dataset samples.
     """  # noqa: E501
-    return from_items(list(dataset))
+    return read_datasource(
```
Comment (amogkam) [inline python/ray/data/read_api.py:2320]: we should only support parallel reads for map-style PyTorch Datasets. For iterable-style pytorch datasets, reading in parallel will read the entire Dataset N times. For this case, we should only read in 1 read task.

File: python/ray/data/read_api.py [hunk 9]
```diff
@@ -2282,6 +2283,9 @@ def from_tf(
 @PublicAPI
 def from_torch(
     dataset: "torch.utils.data.Dataset",
+    parallelism: int = -1,
```
Comment (amogkam) [inline python/ray/data/read_api.py:2286]: We also need to update this documentation: https://docs.ray.io/en/master/data/working-with-pytorch.html#migrating-from-pytorch-datasets-and-dataloaders

File: python/ray/data/read_api.py [hunk 10]
```diff
@@ -2282,6 +2283,9 @@ def from_tf(
...
+    random_split: bool = False,
+    batch_size: int = 32,
 ) -> MaterializedDataset:
     """Create a :class:`~ray.data.Dataset` from a
     `Torch Dataset <https://pytorch.org/docs/stable/data.html#torch.utils.data.Dataset/>`_.
```
Comment (amogkam) [inline python/ray/data/read_api.py:2291]: We need to update the docstring to reflect new changes. Things like this function being inefficient or executing slowly may not be applicable with the new changes.

File: python/ray/data/read_api.py (hunk 8 above)
Comment (amogkam) [inline python/ray/data/read_api.py:2320]: we also should clarify that by default this requires the underlying files to be accessible from every node in the cluster.

we should also provide a way for users to call `from_torch` if only the head node has access to the underlying files.

File: python/ray/data/datasource/torch_datasource.py (hunk 1 above)
Comment (amogkam) [reply python/ray/data/datasource/torch_datasource.py:31]: +1... I don't think there's a good way to calculate dataset size right now, just as with huggingface datasource

File: python/ray/data/datasource/torch_datasource.py (hunk 6 above)
Comment (Zandew) [reply python/ray/data/datasource/torch_datasource.py:81]: If we were to just access the dataset directly, would there be any point in still batching it?

File: python/ray/data/read_api.py (hunk 8 above)
Comment (Zandew) [reply python/ray/data/read_api.py:2320]: Could you clarify what this means? When are files not accessible to certain nodes in the cluster?

File: python/ray/data/read_api.py (hunk 7 above)
Comment (Zandew) [reply python/ray/data/read_api.py:2288]: random_split doesn't actually shuffle the dataset, it just partitions the dataset randomly. The order within those partitions remain the same in the final dataset. @scottjlee recommended changing the name to `shuffle_blocks`

File: python/ray/data/datasource/torch_datasource.py (hunk 6 above)
Comment (amogkam) [reply python/ray/data/datasource/torch_datasource.py:81]: Yep exactly, we don't need to batch it and can remove the DataLoader

File: python/ray/data/read_api.py (hunk 7 above)
Comment (amogkam) [reply python/ray/data/read_api.py:2288]: We should shuffle rows instead of shuffling blocks.

Create a list of integers from 0 to len(dataset).
Shuffle this list of integers
Then partition this list of integers evenly across all the read tasks

File: python/ray/data/read_api.py (hunk 8 above)
Comment (amogkam) [reply python/ray/data/read_api.py:2320]: If there are files that are only the head node filesystem, then they cannot be accessed by worker nodes. The recommended approach should be to store the underlying files in either a shared filesystem or in cloud storage.

File: python/ray/data/read_api.py (hunk 7 above)
Comment (Zandew) [reply python/ray/data/read_api.py:2288]: Actually, I read wrong, random_split does not maintain relative order inside splits, so this is a true shuffle.

File: python/ray/data/read_api.py (hunk 8 above)
Comment (Zandew) [reply python/ray/data/read_api.py:2334]: Do you know if this is being done for any other data sources?

File: python/ray/data/read_api.py [hunk 11]
```diff
@@ -2282,21 +2283,24 @@ def from_tf(
...
-    This function is inefficient. Use it to read small datasets or prototype.
-
     .. warning::
         If your dataset is large, this function may execute slowly or raise an
         out-of-memory error. To avoid issues, read the underyling data with a function
         like :meth:`~ray.data.read_images`.
 
     .. note::
-        This function isn't parallelized. It loads the entire dataset into the head
+        For iterable-style datasets, This function is not parallelized. It loads the entire dataset into the head
```
Comment (scottjlee) [inline python/ray/data/read_api.py:2298]: nit:
```suggestion
        For iterable-style datasets, this function is not parallelized. It loads the entire dataset into the head
```

File: python/ray/data/datasource/torch_datasource.py (hunk 6 above)
Comment (Zandew) [reply python/ray/data/datasource/torch_datasource.py]: I tried running `test_from_torch_parallel` with DataLoader and it takes around 18s. Without DataLoader and batching it takes around 66s. It looks like there's an optimization performed by the DataLoader?

Comment (Zandew) [reply python/ray/data/datasource/torch_datasource.py]: I also tried manually batching without the DataLoader, it looks like the speedup comes from batching and not the DataLoader

File: python/ray/data/read_api.py (hunk 11 above)
Comment (scottjlee) [inline python/ray/data/read_api.py:2298]: let's also clarify that this implementation is now a streamed implementation, so can read larger than memory datasets

File: python/ray/data/read_api.py [hunk 12]
```diff
@@ -2309,11 +2313,30 @@ def from_torch(
...
+        shuffle: If True, the dataset will be randomly partitioned for a parallel read. Otherwise, the dataset will be partitioned into contiguous blocks.
 
     Returns:
         A :class:`MaterializedDataset` containing the Torch dataset samples.
     """  # noqa: E501
-    return from_items(list(dataset))
+    import torch
+
+    # There is no efficient way to read a subset of an iterable-style dataset.
+    # Each task will read all elements up until the last in it's subset.
```
Comment (scottjlee) [inline python/ray/data/read_api.py:2331]: nit
```suggestion
    # Each task will read all elements up until the last in its subset.
```

File: python/ray/data/datasource/torch_datasource.py [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+        return read_tasks
+
+    def estimate_inmemory_data_size(self):
+        return None
+
+
+def _read_subset(subset: "torch.utils.data.Subset"):
+    import torch
+
+    loader = torch.utils.data.DataLoader(subset, collate_fn=lambda x: x, batch_size=32)
```
Comment (scottjlee) [inline python/ray/data/datasource/torch_datasource.py:88]: i guess we should add this `batch_size` as a parameter that we can configure from the `read_api` (e.g. using `**read_args`), since users may need to tune this value.

Comment (amogkam) [reply python/ray/data/datasource/torch_datasource.py:88]: we don't need to use a dataloader here. let's just index the dataset directly?

Comment (Zandew) [reply python/ray/data/datasource/torch_datasource.py:88]: 👍

File: python/ray/data/read_api.py [hunk 14]
```diff
@@ -2282,21 +2283,28 @@ def from_tf(
...
     .. note::
-        This function isn't parallelized. It loads the entire dataset into the head
+        This function uses a streaming implementation. The dataset does not need to fit completely into memory.
+
+    .. note::
+        For iterable-style datasets, this function is not parallelized. It loads the entire dataset into the head
         node's memory before moving the data to the distributed object store.
 
+    .. note::
+        Parallel read dispatches read tasks to worker nodes. The underlying files of the dataset must be accesible from every node in the cluster.
```
Comment (raulchen) [inline python/ray/data/read_api.py:2306]: I think it's pretty common that the data is only accessible on the local, right? E.g. when setting `download=True`. 
I'm not sure if torch data has an API to tell if the data is local only. If not, we should add a parameter that allows users to force running the read tasks on local node.

File: python/ray/data/read_api.py [hunk 15]
```diff
@@ -2282,21 +2283,28 @@ def from_tf(
 @PublicAPI
 def from_torch(
     dataset: "torch.utils.data.Dataset",
+    parallelism: int = -1,
+    shuffle: bool = False,
+    batch_size: int = 32,
```
Comment (amogkam) [inline python/ray/data/read_api.py:2288]: this can be removed?

File: python/ray/data/read_api.py [hunk 16]
```diff
@@ -2282,21 +2283,28 @@ def from_tf(
...
         If your dataset is large, this function may execute slowly or raise an
         out-of-memory error. To avoid issues, read the underyling data with a function
         like :meth:`~ray.data.read_images`.
 
     .. note::
-        This function isn't parallelized. It loads the entire dataset into the head
+        This function uses a streaming implementation. The dataset does not need to fit completely into memory.
+
+    .. note::
+        For iterable-style datasets, this function is not parallelized. It loads the entire dataset into the head
```
Comment (amogkam) [inline python/ray/data/read_api.py:2302]: for iterable datasets we should still do it in a streaming fashion right? it just won't be parallelized.

File: python/ray/data/read_api.py (hunk 14 above)
Comment (amogkam) [inline python/ray/data/read_api.py:2306]: should we add an argument to work with the case where the files are only on the head node?

Comment (amogkam) [inline python/ray/data/read_api.py:2306]: let's move this note up higher

Comment (amogkam) [reply python/ray/data/read_api.py:2306]:
```suggestion
        Parallel read dispatches read tasks to worker nodes. The underlying files of the dataset must be accesible from every node in the cluster (for example the files are stored on cloud storage or a shared filesystem)
```

File: python/ray/data/read_api.py [hunk 17]
```diff
@@ -2282,21 +2283,28 @@ def from_tf(
...
-    This function is inefficient. Use it to read small datasets or prototype.
-
     .. warning::
         If your dataset is large, this function may execute slowly or raise an
         out-of-memory error. To avoid issues, read the underyling data with a function
         like :meth:`~ray.data.read_images`.
 
     .. note::
-        This function isn't parallelized. It loads the entire dataset into the head
+        This function uses a streaming implementation. The dataset does not need to fit completely into memory.
```
Comment (amogkam) [inline python/ray/data/read_api.py:2299]: the warning above this is no longer applicable

File: python/ray/data/datasource/torch_datasource.py [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+
+        return read_tasks
+
+    def estimate_inmemory_data_size(self):
+        return None
+
+
+def _read_subset(subset: "torch.utils.data.Subset", batch_size: int):
+    batch = []
+    for item in subset:
```
Comment (amogkam) [inline python/ray/data/datasource/torch_datasource.py:92]: iterating through the torch dataset will output individual items, not full batches

File: python/ray/data/read_api.py (hunk 16 above)
Comment (Zandew) [reply python/ray/data/read_api.py:2300]: Yup, it should still be streamed

Comment (amogkam) [reply python/ray/data/read_api.py:2300]: The second sentence is not correct then?

Comment (Zandew) [reply python/ray/data/read_api.py:2300]: Yeah that's true, let me remove that

File: python/ray/data/datasource/torch_datasource.py [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+
+TORCH_DATASOURCE_READER_BATCH_SIZE = 32
+
+
+@DeveloperAPI
+class TorchDatasource(Datasource):
+    """Torch datasource, for reading from `map-style
+    Torch datasets <https://pytorch.org/docs/stable/data.html#map-style-datasets/>`_.
+    This datasource implements a parallel read that partitions the dataset based on
+    input parallelism and creates read tasks for each partitions.
```
Comment (raulchen) [inline python/ray/data/datasource/torch_datasource.py:21]: also update this docstring to reflect the latest implementaion

**Review by raulchen — APPROVED** [review]
